File: cpp/src/gameinfo.cc
// gameinfo.cc
//
// Copyright (C) 2011 - TANIGUCHI Takaki
//
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 2 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program. If not, see <http://www.gnu.org/licenses/>.

#include "defs.h"
#include "gameinfo.h"
#include "ball.h"
#include <glibmm/random.h>
#include <cmath>

GameInfo::GameInfo() :
slime_blue(BLUE),
slime_red(RED)
{
	init();
}

void
GameInfo::init()
{
	ball.init(BLUE);
	slime_blue.init();
	slime_red.init();
	ball_owner = BLUE;
	ball_count = 1;
	wait_count = 0;
	score_blue = SCORE_INIT;
	score_red  = SCORE_INIT;
	state = GS_SERVICE_BY_BLUE;
}

void 
GameInfo::score_move(TEAM win)
{
	if (ball_count > 0 ) {
		if (wait_count > 0) {
			--wait_count;
		} else  {
			--ball_count;
			if (ball_count == 0) {
				wait_count = 50; // XXX
			} else {
				wait_count = 7; // XXX
			}
			if (win == BLUE) {
				--score_red;
				if (score_red <= 0) {
					state = GS_WON_BY_BLUE;
				}
			} else {
				--score_blue;
				if (score_blue <= 0) {
					state = GS_WON_BY_RED;
				}
			}			
		}
	} else {
		if (wait_count > 0 ){
			--wait_count;
		} else {
			if (win == BLUE) {
				state = GS_SERVICE_BY_BLUE;
			} else {
				state = GS_SERVICE_BY_RED;
			}
		}
	}
};

void 
GameInfo::serve_set(TEAM win)
{
	if (win == BLUE) {
		ball.init(BLUE);
	} else {
		ball.init(RED);
	}
	slime_blue.init();
	slime_red.init();
	Glib::Rand r;
	penalty_y = r.get_double_range(120, WIN_HEIGHT - 40); // XXX	
	ball_count = 1;
	wait_count = 0;
	state = GS_PLAY;
}

typedef
struct {
	double x;
	double y;
} Vector;

static double
vector_inner_product(Vector a, Vector b)
{
	return (a.x * b.x + a.y * b.y);
};
static double
vector_outer_product(Vector a, Vector b)
{
	return (a.x * b.y - b.x * a.y);
};

static double
vector_norm(Vector a)
{
	return sqrt(a.x*a.x+a.y*a.y);
};

void
GameInfo::collision_ball_slime(TEAM t){
	Ball &b = ball;
	Slime *s;
	if (t == BLUE ){
		s = &slime_blue;
	} else {
		s = &slime_red;
	}
	Vector v_s = {b.get_vx(), b.get_vy()};
	Vector v_a = {s->get_x() - b.get_x(), WIN_HEIGHT-b.get_y()};
	Vector v_b = {s->get_x() - (b.get_x() + b.get_vx()), WIN_HEIGHT - (b.get_y() + b.get_vy())};
	double c = vector_inner_product (v_a, v_s) * 
		vector_inner_product (v_b, v_s) ;
	double d = fabs(vector_outer_product (v_s, v_a))/ vector_norm(v_s);
	double r = SLIME_RADIUS + BALL_RADIUS;

	if ( d <= r && ( c<=0 || r > vector_norm (v_a) || r > vector_norm (v_b))){
		double dx_0 = b.get_x() - s->get_x();
		double dy_0 = b.get_y() - WIN_HEIGHT;
		double dist_0 = sqrt(pow(dx_0, 2) + pow(dy_0, 2));
		double angle = acos (dx_0/dist_0);
		double cos2x = cos(2*angle);
		double sin2x = sin(2*angle);
		double vx = - b.get_vx() * cos2x + b.get_vy() * sin2x;
		double vy =   b.get_vx() * sin2x + b.get_vy() * cos2x;		
		b.set_vx(vx);
		b.set_vy(vy);

		ball_owner = t;
		ball_count++;
	}
}

void
GameInfo::collision_penalty ()
{
	Ball &b = ball;
	if (sqrt(pow(b.get_x()-WIN_WIDTH/2,2)+pow(b.get_y() - penalty_y,2)) < 
	    PENALTY_RADIUS + BALL_RADIUS)
	{
		--wait_count;
		if (ball_owner == BLUE ) {
			if (score_blue > 1 && wait_count <= 0) {
				--score_blue;				
				wait_count = 4; // XXX
			}				
		} else {
			if (score_red > 1 && wait_count <= 0) {
				--score_red;
				wait_count = 4; // XXX
			}
		}
	}
}

File: java/src/GameInfo.java
import java.util.Random;

public class GameInfo {
	enum GameState {
		GS_INIT, GS_PLAY, GS_GOT_BY_BLUE, GS_GOT_BY_RED, GS_SERVICE_BY_BLUE, GS_SERVICE_BY_RED, GS_WON_BY_BLUE, GS_WON_BY_RED,
	}

	private static final int SCORE_INIT = 50;;

	final private Slime slime_blue = new Slime(Slime.TEAM.BLUE);
	public Slime getSlime_blue() {
		return slime_blue;
	}

	public Slime getSlime_red() {
		return slime_red;
	}

	final private Slime slime_red = new Slime(Slime.TEAM.RED);
	final private Ball ball = new Ball();

	public Ball getBall() {
		return ball;
	}

	private GameState state;

	public GameState getState() {
		return state;
	}

	public void setState(GameState state) {
		this.state = state;
	}

	private int score_blue = SCORE_INIT;
	public int getScore_blue() {
		return score_blue;
	}

	public int getScore_red() {
		return score_red;
	}

	private int score_red = SCORE_INIT;
	private int ball_count = 1;
	public int getBall_count() {
		return ball_count;
	}

	private Slime.TEAM ball_owner = Slime.TEAM.BLUE;
	private double penalty_y;
	private int wait_count = 0;

	public GameInfo() {
		ball.init(Slime.TEAM.BLUE);
		slime_blue.init();
		slime_red.init();
		ball_count = 1;
		// state = GameState.GS_SERVICE_BY_BLUE; // XXX GS_INIT??
		state = GameState.GS_INIT;
	}

	public void score_move(Slime.TEAM win) {
		if (ball_count > 0) {
			if (wait_count > 0) {
				--wait_count;
			} else {
				--ball_count;
				if (ball_count == 0) {
					wait_count = 50; // XXX
				} else {
					wait_count = 7; // XXX
				}
				if (win == Slime.TEAM.BLUE) {
					--score_red;
					if (score_red <= 0) {
						state = GameState.GS_WON_BY_BLUE;
					}
				} else {
					--score_blue;
					if (score_blue <= 0) {
						state = GameState.GS_WON_BY_RED;
					}
				}
			}
		} else {
			if (wait_count > 0) {
				--wait_count;
			} else {
				if (win == Slime.TEAM.BLUE) {
					state = GameState.GS_SERVICE_BY_BLUE;
				} else {
					state = GameState.GS_SERVICE_BY_RED;
				}
			}
		}
	};

	public void serve_set(Slime.TEAM win) {
		if (win == Slime.TEAM.BLUE) {
			ball.init(Slime.TEAM.BLUE);
		} else {
			ball.init(Slime.TEAM.RED);
		}
		slime_blue.init();
		slime_red.init();
		Random r = new Random();

		penalty_y = ((GameWin.WIN_HEIGHT - 40) - 120) * r.nextDouble() + 120;

		ball_count = 1;
		wait_count = 0;
		state = GameState.GS_PLAY;
	}

	private double inner_product(double a[], double b[]) {
		return (a[0] * b[0] + a[1] * b[1]);
	};

	double outer_product(double a[], double b[]) {
		return (a[0] * b[1] - b[0] * a[1]);
	};

	double norm(double a[]) {
		return Math.sqrt(a[0] * a[0] + a[1] * a[1]);
	};

	void collision_ball_slime(Slime.TEAM t) {
		final Ball b = ball;
		final Slime s;
		if (t == Slime.TEAM.BLUE) {
			s = slime_blue;
		} else {
			s = slime_red;
		}
		double v_s[] = { b.getVx(), b.getVy() };
		double v_a[] = { s.getX() - b.getX(), GameWin.WIN_HEIGHT - b.getY() };
		double v_b[] = { s.getX() - (b.getX() + b.getVx()),
				GameWin.WIN_HEIGHT - (b.getY() + b.getVy()) };
		double c = inner_product(v_a, v_s) * inner_product(v_b, v_s);
		double d = Math.abs(outer_product(v_s, v_a)) / norm(v_s);
		double r = Slime.RADIUS + Ball.RADIUS;

		if (d <= r && (c <= 0 || r > norm(v_a) || r > norm(v_b))) {
			double dx_0 = b.getX() - s.getX();
			double dy_0 = b.getY() - GameWin.WIN_HEIGHT;
			double[] v = { dx_0, dy_0 };
			double dist_0 = norm(v);
			double angle = Math.acos(dx_0 / dist_0);
			double cos2x = Math.cos(2 * angle);
			double sin2x = Math.sin(2 * angle);
			double vx = -b.getVx() * cos2x + b.getVy() * sin2x;
			double vy = b.getVx() * sin2x + b.getVy() * cos2x;
			b.setVx(vx);
			b.setVy(vy);

			ball_owner = t;
			ball_count++;
		}
	}

	public void collision_penalty() {
		final Ball b = ball;
		double v[] = { b.getX() - GameWin.WIN_WIDTH / 2, b.getY() - penalty_y };

		if (norm(v) < GameWin.PENALTY_RADIUS + Ball.RADIUS) {
			--wait_count;
			if (ball_owner == Slime.TEAM.BLUE) {
				if (score_blue > 1 && wait_count <= 0) {
					--score_blue;
					wait_count = 4; // XXX
				}
			} else {
				if (score_red > 1 && wait_count <= 0) {
					--score_red;
					wait_count = 4; // XXX
				}
			}
		}
	}

	public double getPenalty_y() {
		return penalty_y;
	}

}
